chirp/drivers/baofeng_uv17.py

```diff
@@ -0,0 +1,423 @@
+        basic.append(rs)
+
+        rs = RadioSetting("settings.scanmode", "Scan Mode",
+                          RadioSettingValueList(
+                              LIST_SCANMODE,
+                              LIST_SCANMODE[_mem.settings.scanmode]))
+        basic.append(rs)
+
+        rs = RadioSetting("settings.dtmfst", "DTMF Sidetone",
+                          RadioSettingValueList(LIST_DTMFST, LIST_DTMFST[
+                              _mem.settings.dtmfst]))
+        basic.append(rs)
+
+        rs = RadioSetting("settings.fmenable", "Enable FM radio",
+                          RadioSettingValueBoolean(_mem.settings.fmenable))
+        basic.append(rs)
+
+        self.get_settings_common_dtmf(dtmfe)
+
+        return top
+
+    def decode_tone(self, val):
+        pol = "N"
+        mode = ""
+        if val in [0, 0xFFFF]:
+            xval = 0
+        elif (val & 0x8000) > 0:
+            mode = "DTCS"
+            xval = (val & 0x0f) + (val >> 4 & 0xf)\
+                * 10 + (val >> 8 & 0xf) * 100
+            if (val & 0xC000) == 0xC000:
+                pol = "R"
+        else:
+            mode = "Tone"
+            xval = int((val & 0x0f) + (val >> 4 & 0xf) * 10 +
+                       (val >> 8 & 0xf) * 100 + (val >> 12 & 0xf)
+                       * 1000) / 10.0
+
+        return mode, xval, pol
+
+    def get_memory(self, number):
+        offset = 0
+        # Skip 16 bytes at memory block boundary
+        # This is because these 16 bytes contain memory page numbers
+        if number >= 252:
+            offset += 1
+        if number >= 507:
+            offset += 1
+        if number >= 762:
+            offset += 1
+
+        _mem = self._memobj.memory[number + offset]
+        _nam = self._memobj.names[number]
+
+        mem = chirp_common.Memory()
+        mem.number = number
+
+        self.get_memory_common(_mem, _nam.name, mem)
+
+        return mem
+
+    def encode_tone(self, memtone, mode, tone, pol):
+        if mode == "Tone":
+            xtone = str(int(tone * 10)).rjust(4, '0')
+            memtone.set_value((int(xtone[0]) << 12) + (int(xtone[1]) << 8) +
+                              (int(xtone[2]) << 4) + int(xtone[3]))
+        elif mode == "TSQL":
+            xtone = str(int(tone * 10)).rjust(4, '0')
+            memtone.set_value((int(tone[0]) << 12) + (int(xtone[1]) << 8) +
+                              (int(xtone[2]) << 4) + int(xtone[3]))
+        elif mode == "DTCS":
+            xtone = str(int(tone)).rjust(4, '0')
+            memtone.set_value((0x8000 + (int(xtone[0]) << 12) +
+                               (int(xtone[1]) << 8) + (int(xtone[2]) << 4) +
+                               int(xtone[3])))
+        else:
+            memtone.set_value(0)
+
+        if mode == "DTCS" and pol == "R":
+            memtone.set_value(memtone + 0x4000)
+
+    def set_memory(self, mem):
+        offset = 0
+        # skip 16 bytes at memory block boundary
```

Comment:
Does this comment refer to the conditional offsets below?

You can use `#skip` in your memory format to move forward a specific amount, if it helps. When I've written drivers that don't store memories all linearly, I generally try to define the format declaratively instead of doing stuff like this. Something like:
```
struct mem lowmems[252];
#skip 0x0x10;
struct mem highmems[506];
// ... etc
```

Your comment in `get_memory()` mentions that there's a 16-byte "page number". Not really sure what that means but you can also do something like this:
```
struct {
  struct mem memories[252];
  u8 page_stuff[16];
} mem_page[3];
```

The reason to define it like this instead of just papering over it the way you are doing here, is that these "page number" structures will be represented in the browser as memories with non-sensical data in them, which is not great.

At the very least, you should probably make this logic a helper so you don't have to repeat it in `get_memory()` and `set_memory()`, especially if you ever have to tweak it.

Oh and this makes me realize - you didn't implement `get_raw_memory()` here or for the pro one. You should do that, because it makes it *much* easier to iterate and fix things.

Reply:
Fixed. Added get_raw_memory and removed the offset stuff..